pubsub/v2/shutdown.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pubsub
+
+import "time"
+
+// ShutdownOptions configures the shutdown behavior of the subscriber.
+// If not specified, the behavior will default to indefinite processing,
+// that is graceful shutdown with no timeout.
+type ShutdownOptions struct {
+	// Timeout specifies the time the subscriber should wait
+	// to shutdown before killing the process.
+	// In nack mode, this specifies how long to wait
+	// for messages to be nacked after shutdown is initiated.
+	//
+	// Set to zero to immediately shutdown.
+	// Set to a negative value to disable timeout.
+	Timeout time.Duration
```

Comment:
Do we need to disambiguate the default value here?  This isn't a proto where you have field presence that can be checked.  Consider the "wait indefinitely" and "don't wait die immediately" user choices.

Reply:
Mostly the inversion of timeout behavior from the unspecified ShutdownOptions option to the empty option is what make me itchy here.  In the first we wait indefinitely, in the second we die immediately.

Reply:
The "default" value for Timeout is interesting, because it only is relevant when user sets a ShutdownOption.Behavior but not Timeout.

I added a comment warning that if you don't specify both fields when setting `ShutdownOptions`, you will get the forceful shutdown.

There wasn't really a clean choice here, other than maybe to use another optional/pointer value here. In order to get the default/zero value as "wait indefinitely", we have to change the semantics of Timeout=0 to be wait indefinitely, where Timeout=-1 means shutdown immediately. To me, this is more confusing (especially in Pub/Sub where `modack(0)` equals nack / immediately redeliver) and why we decided to implement this way.

Reply:
I chatted with @shollyman and agreed that while it's unfortunate the nil and zero values of `ShutdownOptions` do different things, if it's sufficiently documented, we can proceed without a "Legacy" mode. As mentioned earlier, the presence of "Legacy" means that users won't be able to do "quit immediately" by just setting

```&pubsub.ShutdownOptions{Timeout: 0}`

They would have to specify one of the other non-legacy options (that they're not using)

```&pubsub.ShutdownOptions{
    Timeout: 0,
    Behavior: WaitForProcessing // this has no semantic meaning here
}`